rust/system/src/types.rs

```diff
@@ -92,26 +95,55 @@ where
 
 /// A thin wrapper over a join handle that will panic if it is consumed more than once.
 #[derive(Debug, Clone)]
-pub(super) struct ConsumableJoinHandle {
-    handle: Arc<Mutex<Option<tokio::task::JoinHandle<()>>>>,
+pub(super) enum ConsumableJoinHandle {
+    TokioTask(Arc<Mutex<Option<tokio::task::JoinHandle<()>>>>),
+    Thread(Arc<Mutex<Option<std::thread::JoinHandle<()>>>>),
+}
+
+#[derive(Debug, Error)]
+pub enum ConsumeJoinHandleError {
+    #[error("Tokio task failed: {0}")]
+    TokioTaskFailed(#[from] tokio::task::JoinError),
+    #[error("Thread panicked")]
+    ThreadPanicked(Box<dyn std::any::Any + Send + 'static>),
 }
 
 impl ConsumableJoinHandle {
-    pub(super) fn new(handle: tokio::task::JoinHandle<()>) -> Self {
-        ConsumableJoinHandle {
-            handle: Arc::new(Mutex::new(Some(handle))),
-        }
+    pub(super) fn from_tokio_task_handle(handle: tokio::task::JoinHandle<()>) -> Self {
+        ConsumableJoinHandle::TokioTask(Arc::new(Mutex::new(Some(handle))))
+    }
+
+    pub(super) fn from_thread_handle(handle: std::thread::JoinHandle<()>) -> Self {
+        ConsumableJoinHandle::Thread(Arc::new(Mutex::new(Some(handle))))
     }
 
-    async fn consume(&mut self) -> Result<(), JoinError> {
-        let handle = { self.handle.lock().take() };
-        match handle {
-            Some(handle) => {
-                handle.await?;
-                Ok(())
+    async fn consume(&mut self) -> Result<(), ConsumeJoinHandleError> {
+        match self {
+            ConsumableJoinHandle::TokioTask(handle) => {
+                let handle = { handle.lock().take() };
+                match handle {
+                    Some(handle) => {
+                        handle.await?;
+                        Ok(())
+                    }
+                    None => {
+                        panic!("Join handle already consumed");
+                    }
+                }
             }
-            None => {
-                panic!("Join handle already consumed");
+            ConsumableJoinHandle::Thread(handle) => {
+                let handle = { handle.lock().take() };
+                match handle {
+                    Some(handle) => {
+                        handle
```

Comment:
Will this block the runtime?

Reply:
I.e if a single threaded runtime called join here, the runtime would make no progress until this thread joined. So if that thread was somehow dependent on a completion from the runtime to join, this would deadlock.